Support for call hierarchy

It seems that [LSP 3.16](https://microsoft.github.io/language-server-protocol/specification#version_3_16_0) supports call hierarchy query. Would it be possible to support this in kak-lsp?

That's an interesting feature, very powerful in theory.
See #560 for a basic implementation.

I'm not sure what the UI should look like.

Here's the buffer for outgoing calls of `InitWithCustomCA()` from https://github.com/zaquestion/lab

```
internal/gitlab/gitlab.go:107:6: InitWithCustomCA
  /usr/lib/go/src/io/ioutil/ioutil.go:36:6: ReadFile
    internal/gitlab/gitlab.go:108:24: 	caCert, err := ioutil.ReadFile(caFile)
  /usr/lib/go/src/crypto/x509/cert_pool.go:105:6: SystemCertPool
    internal/gitlab/gitlab.go:113:26: 	caCertPool, err := x509.SystemCertPool()
  /usr/lib/go/src/crypto/x509/cert_pool.go:206:20: AppendCertsFromPEM
    internal/gitlab/gitlab.go:117:13: 	caCertPool.AppendCertsFromPEM(caCert)
  /home/johannes/git/go-gitlab/gitlab.go:209:6: NewClient
    internal/gitlab/gitlab.go:138:18: 	lab, _ = gitlab.NewClient(token, gitlab.WithHTTPClient(httpClient), gitlab.WithBaseURL(host+"/api/v4"))
  /home/johannes/git/go-gitlab/client_options.go:79:6: WithHTTPClient
    internal/gitlab/gitlab.go:138:42: 	lab, _ = gitlab.NewClient(token, gitlab.WithHTTPClient(httpClient), gitlab.WithBaseURL(host+"/api/v4"))
  /home/johannes/git/go-gitlab/client_options.go:29:6: WithBaseURL
    internal/gitlab/gitlab.go:138:77: 	lab, _ = gitlab.NewClient(token, gitlab.WithHTTPClient(httpClient), gitlab.WithBaseURL(host+"/api/v4"))
```

1. The first line is the reference (in this case the caller)
2. Lines indented with two spaces are the destinations of one or more calls in that function
3. Lines indented with four spaces are their call sites

Intuitively I'd swap indentation of 2 and 3, or maybe not even show 2.

--- a/README.asciidoc
+++ b/README.asciidoc
@@ -142,6 +142,8 @@ Either way you get:
 ** `\*symbols*` buffer has filetype `lsp-goto` so you can press `<ret>` on a line or use the `lsp-jump` command
 * `lsp-diagnostics` command to list project-wide diagnostics (current buffer determines project and language to collect diagnostics for)
 ** `\*diagnostics*` buffer has filetype `lsp-goto` so you can press `<ret>` on a line or use the `lsp-jump` command
+* `lsp-incoming-calls` and `lsp-outgoing-calls` commands to list callers and callees of the function at the cursor.
+** `\*callers*` and `\*callees*` buffers have filetype `lsp-goto` so you can press `<ret>` on a line or use the `lsp-jump` command
 * inline diagnostics highlighting using the `DiagnosticError`, `DiagnosticHint`, `DiagnosticInfo` and `DiagnosticWarning` faces; can be disabled with `lsp-inline-diagnostics-disable` command
 * flags in the left margin on lines with errors or other diagnostics; can be disabled with `lsp-diagnostic-lines-disable` command
 * `lsp-formatting` command to format current buffer, according to the `tabstop` and `lsp_insert_spaces` options
@@ -184,6 +186,8 @@ map global user l %{: enter-user-mode lsp<ret>} -docstring "LSP mode"
 | f | lsp-formatting
 | h | lsp-hover
 | i | lsp-implementation
+| j | lsp-outgoing-calls
+| k | lsp-incoming-calls
 | r | lsp-references
 | R | lsp-rename-prompt
 | s | lsp-signature-help

--- a/rc/lsp.kak
+++ b/rc/lsp.kak
@@ -780,6 +780,29 @@ cat ${pipe} | tee /tmp/pipe
 rm -rf ${tmp}
 }}
 
+define-command lsp-incoming-calls -docstring "Open buffer with calls to the function at the main cursor position" %{
+    lsp-call-hierarchy-request true
+}
+
+define-command lsp-outgoing-calls -docstring "Open buffer with calls by the function at the main cursor position" %{
+    lsp-call-hierarchy-request false
+}
+
+define-command -hidden lsp-call-hierarchy-request -params 1 %{
+    nop %sh{ (printf '
+session   = "%s"
+client    = "%s"
+buffile   = "%s"
+filetype  = "%s"
+version   = %d
+method    = "textDocument/prepareCallHierarchy"
+[params]
+position.line = %d
+position.column = %d
+incomingOrOutgoing = %s
+' "${kak_session}" "${kak_client}" "${kak_buffile}" "${kak_opt_filetype}" "${kak_timestamp}" "${kak_cursor_line}" "${kak_cursor_column}" "$1" | eval ${kak_opt_lsp_cmd} --request) > /dev/null 2>&1 < /dev/null & }
+}
+
 # CCLS Extension
 
 define-command ccls-navigate -docstring "Navigate C/C++/ObjectiveC file" -params 1 %{
@@ -1040,6 +1063,14 @@ define-command -hidden lsp-show-goto-choices -params 2 -docstring "Render goto c
 
 define-command -hidden lsp-show-document-symbol -params 2 -docstring "Render document symbols" %{
     lsp-show-goto-buffer *goto* %arg{@}
+}
+
+define-command -hidden lsp-show-incoming-calls -params 2 -docstring "Render callers" %{
+    lsp-show-goto-buffer *callers* %arg{@}
+}
+
+define-command -hidden lsp-show-outgoing-calls -params 2 -docstring "Render callees" %{
+    lsp-show-goto-buffer *callees* %arg{@}
 }
 
 define-command lsp-next-location -params 1 -docstring %{
@@ -1378,6 +1409,8 @@ map global lsp e '<esc>: lsp-diagnostics<ret>'            -docstring 'list proje
 map global lsp f '<esc>: lsp-formatting<ret>'             -docstring 'format buffer'
 map global lsp h '<esc>: lsp-hover<ret>'                  -docstring 'show info for current position'
 map global lsp i '<esc>: lsp-implementation<ret>'         -docstring 'go to implementation'
+map global lsp j '<esc>: lsp-outgoing-calls<ret>'         -docstring 'list outgoing call for function at cursor'
+map global lsp k '<esc>: lsp-incoming-calls<ret>'         -docstring 'list incoming call for function at cursor'
 map global lsp r '<esc>: lsp-references<ret>'             -docstring 'list symbol references'
 map global lsp R '<esc>: lsp-rename-prompt<ret>'          -docstring 'rename symbol'
 map global lsp s '<esc>: lsp-signature-help<ret>'         -docstring 'show function signature help'

--- a/rc/lsp.kak
+++ b/rc/lsp.kak
@@ -1023,26 +1023,23 @@ define-command -hidden lsp-show-diagnostics -params 2 -docstring "Render diagnos
     }
 }
 
-define-command -hidden lsp-show-goto-choices -params 2 -docstring "Render goto choices" %{
+define-command -hidden lsp-show-goto-buffer -params 3 %{
     evaluate-commands -save-regs '"' -try-client %opt[toolsclient] %{
-        edit! -scratch *goto*
+        edit! -scratch %arg{1}
         set-option buffer filetype lsp-goto
         set-option buffer grep_current_line 0
-        set-option buffer lsp_project_root "%arg{1}/"
-        set-register '"' %arg{2}
+        set-option buffer lsp_project_root "%arg{2}/"
+        set-register '"' %arg{3}
         execute-keys Pgg
     }
 }
 
+define-command -hidden lsp-show-goto-choices -params 2 -docstring "Render goto choices" %{
+    lsp-show-goto-buffer *goto* %arg{@}
+}
+
 define-command -hidden lsp-show-document-symbol -params 2 -docstring "Render document symbols" %{
-    evaluate-commands -save-regs '"' -try-client %opt[toolsclient] %{
-        edit! -scratch *symbols*
-        set-option buffer filetype lsp-goto
-        set-option buffer grep_current_line 0
-        set-option buffer lsp_project_root "%arg{1}/"
-        set-register '"' %arg{2}
-        execute-keys Pgg
-    }
+    lsp-show-goto-buffer *goto* %arg{@}
 }
 
 define-command lsp-next-location -params 1 -docstring %{

--- a/rc/lsp.kak
+++ b/rc/lsp.kak
@@ -100,7 +100,7 @@ Capture groups must be:
     2: line number
     3: optional column
     4: optional message
-} regex lsp_location_format ^\h*([^:\n]+):(\d+)\b(?::(\d+)\b)?(?::([^\n]+))
+} regex lsp_location_format ^\h*\K([^:\n]+):(\d+)\b(?::(\d+)\b)?(?::([^\n]+))
 
 # Callback functions. Override these to tune kak-lsp's behavior.
 

--- a/rc/lsp.kak
+++ b/rc/lsp.kak
@@ -1765,11 +1765,13 @@ define-command -hidden lsp-make-register-relative-to-root %{
 define-command -hidden lsp-jump %{ # from grep.kak
     evaluate-commands -save-regs abc %{ # use evaluate-commands to ensure jumps are collapsed
         try %{
-            execute-keys "<a-x>s%opt{lsp_location_format}<ret>"
-            set-register a "%reg{1}"
-            set-register b "%reg{2}"
-            set-register c "%reg{3}"
-            lsp-make-register-relative-to-root
+            evaluate-commands -draft %{
+                execute-keys "<a-x>s%opt{lsp_location_format}<ret>"
+                set-register a "%reg{1}"
+                set-register b "%reg{2}"
+                set-register c "%reg{3}"
+                lsp-make-register-relative-to-root
+            }
             set-option buffer grep_current_line %val{cursor_line}
             evaluate-commands -try-client %opt{jumpclient} -verbatim -- edit -existing %reg{a} %reg{b} %reg{c}
             try %{ focus %opt{jumpclient} }
@@ -1779,12 +1781,14 @@ define-command -hidden lsp-jump %{ # from grep.kak
 
 define-command -hidden lsp-diagnostics-jump %{ # from make.kak
     evaluate-commands -save-regs abcd %{
-        execute-keys "<a-x>s%opt{lsp_location_format}<ret>"
-        set-register a "%reg{1}"
-        set-register b "%reg{2}"
-        set-register c "%reg{3}"
-        set-register d "%reg{4}"
-        lsp-make-register-relative-to-root
+        evaluate-commands -draft %{
+            execute-keys "<a-x>s%opt{lsp_location_format}<ret>"
+            set-register a "%reg{1}"
+            set-register b "%reg{2}"
+            set-register c "%reg{3}"
+            set-register d "%reg{4}"
+            lsp-make-register-relative-to-root
+        }
         set-option buffer grep_current_line %val{cursor_line}
         lsp-diagnostics-open-error %reg{a} "%reg{b}" "%reg{c}" "%reg{d}"
     }

--- a/src/controller.rs
+++ b/src/controller.rs
@@ -223,6 +223,9 @@ fn dispatch_editor_request(request: EditorRequest, mut ctx: &mut Context) {
         notification::DidChangeConfiguration::METHOD => {
             workspace::did_change_configuration(meta, params, &mut ctx);
         }
+        request::CallHierarchyPrepare::METHOD => {
+            call_hierarchy::call_hierarchy_prepare(meta, params, &mut ctx);
+        }
         request::Completion::METHOD => {
             completion::text_document_completion(meta, params, &mut ctx);
         }

--- a/src/general.rs
+++ b/src/general.rs
@@ -268,7 +268,9 @@ pub fn initialize(root_path: &str, meta: EditorMeta, ctx: &mut Context) {
                     multiline_token_support: None,
                 }),
                 linked_editing_range: None,
-                call_hierarchy: None,
+                call_hierarchy: Some(CallHierarchyClientCapabilities {
+                    dynamic_registration: Some(false),
+                }),
                 moniker: None,
             }),
             window: Some(WindowClientCapabilities {

--- a/src/language_features/call_hierarchy.rs
+++ b/src/language_features/call_hierarchy.rs
@@ -1,0 +1,193 @@
+use crate::context::*;
+use crate::position::*;
+use crate::types::*;
+use crate::util::*;
+use itertools::Itertools;
+use lsp_types::{request::*, *};
+use serde::Deserialize;
+
+pub fn call_hierarchy_prepare(meta: EditorMeta, params: EditorParams, ctx: &mut Context) {
+    let params = CallHierarchyParams::deserialize(params)
+        .expect("Params should follow CallHierarchyParams structure");
+    let position = get_lsp_position(&meta.buffile, &params.position, ctx).unwrap();
+    let uri = Url::from_file_path(&meta.buffile).unwrap();
+    let prepare_params = CallHierarchyPrepareParams {
+        text_document_position_params: TextDocumentPositionParams {
+            text_document: TextDocumentIdentifier::new(uri),
+            position,
+        },
+        work_done_progress_params: WorkDoneProgressParams::default(),
+    };
+
+    ctx.call::<CallHierarchyPrepare, _>(
+        meta,
+        prepare_params,
+        move |ctx: &mut Context, meta, result| {
+            request_call_hierarchy(meta, ctx, params.incoming_or_outgoing, result);
+        },
+    )
+}
+
+fn request_call_hierarchy(
+    meta: EditorMeta,
+    ctx: &mut Context,
+    incoming_or_outgoing: bool,
+    result: Option<Vec<CallHierarchyItem>>,
+) {
+    // TODO Can we get multiple items here?
+    let item = match result.and_then(|r| r.into_iter().next()) {
+        Some(item) => item,
+        None => return,
+    };
+
+    if incoming_or_outgoing {
+        let params = CallHierarchyIncomingCallsParams {
+            item: item.clone(),
+            work_done_progress_params: WorkDoneProgressParams::default(),
+            partial_result_params: PartialResultParams::default(),
+        };
+
+        ctx.call::<CallHierarchyIncomingCalls, _>(
+            meta,
+            params,
+            move |ctx: &mut Context, meta, result| {
+                format_call_hierarchy_calls(meta, ctx, incoming_or_outgoing, &item, &result);
+            },
+        );
+    } else {
+        let params = CallHierarchyOutgoingCallsParams {
+            item: item.clone(),
+            work_done_progress_params: WorkDoneProgressParams::default(),
+            partial_result_params: PartialResultParams::default(),
+        };
+
+        ctx.call::<CallHierarchyOutgoingCalls, _>(
+            meta,
+            params,
+            move |ctx: &mut Context, meta, result| {
+                format_call_hierarchy_calls(meta, ctx, incoming_or_outgoing, &item, &result);
+            },
+        );
+    }
+}
+
+fn format_location(
+    meta: &EditorMeta,
+    ctx: &mut Context,
+    uri: &Url,
+    position: Position,
+    prefix: &str,
+    suffix: &str,
+) -> String {
+    let filename = uri.to_file_path().unwrap();
+    let filename = short_file_path(filename.to_str().unwrap(), &ctx.root_path);
+    let position = get_kakoune_position_with_fallback(&meta.buffile, position, ctx);
+    format!(
+        "{}{}:{}:{}: {}\n",
+        prefix, filename, position.line, position.column, suffix,
+    )
+}
+
+trait CallHierarchyCall<'a> {
+    fn caller_or_callee(&self) -> &CallHierarchyItem;
+    fn caller(&'a self, other: &'a CallHierarchyItem) -> &'a CallHierarchyItem;
+    fn callsites(&self) -> &Vec<Range>;
+}
+
+impl<'a> CallHierarchyCall<'a> for CallHierarchyIncomingCall {
+    fn caller_or_callee(&self) -> &CallHierarchyItem {
+        &self.from
+    }
+    fn caller(&'a self, _callee: &'a CallHierarchyItem) -> &'a CallHierarchyItem {
+        &self.from
+    }
+    fn callsites(&self) -> &Vec<Range> {
+        &self.from_ranges
+    }
+}
+
+impl<'a> CallHierarchyCall<'a> for CallHierarchyOutgoingCall {
+    fn caller_or_callee(&self) -> &CallHierarchyItem {
+        &self.to
+    }
+    fn caller(&'a self, caller: &'a CallHierarchyItem) -> &'a CallHierarchyItem {
+        caller
+    }
+    fn callsites(&self) -> &Vec<Range> {
+        &self.from_ranges
+    }
+}
+
+fn format_call_hierarchy_calls<'a>(
+    meta: EditorMeta,
+    ctx: &mut Context,
+    incoming_or_outgoing: bool,
+    item: &'a CallHierarchyItem,
+    result: &'a Option<Vec<impl CallHierarchyCall<'a>>>,
+) {
+    let result = match result {
+        Some(result) => result,
+        None => return,
+    };
+
+    let first_line_suffix = format!(
+        "{} - list of {}",
+        &item.name,
+        if incoming_or_outgoing {
+            "callers"
+        } else {
+            "callees"
+        },
+    );
+
+    let contents = format_location(
+        &meta,
+        ctx,
+        &item.uri,
+        item.range.start,
+        "",
+        &first_line_suffix,
+    ) + &result
+        .iter()
+        .map(|call| {
+            let caller = call.caller(item);
+            let callsite_filename = caller.uri.to_file_path().unwrap();
+            let caller_or_calle = call.caller_or_callee();
+
+            format_location(
+                &meta,
+                ctx,
+                &caller_or_calle.uri,
+                caller_or_calle.range.start,
+                "  ",
+                &caller_or_calle.name,
+            ) + &call
+                .callsites()
+                .iter()
+                .map(|range| {
+                    let line = get_file_contents(callsite_filename.to_str().unwrap(), ctx)
+                        .map(|text| text.line(range.start.line as usize).to_string())
+                        .unwrap_or_default();
+                    let line = line
+                        .strip_suffix("\r\n")
+                        .or_else(|| line.strip_suffix('\n'))
+                        .unwrap_or(&line);
+                    format_location(&meta, ctx, &caller.uri, range.start, "    ", line)
+                })
+                .join("")
+        })
+        .join("");
+
+    let command = if incoming_or_outgoing {
+        "lsp-show-incoming-calls"
+    } else {
+        "lsp-show-outgoing-calls"
+    };
+    let command = format!(
+        "{} {} {}",
+        command,
+        editor_quote(&ctx.root_path),
+        editor_quote(&contents),
+    );
+    ctx.exec(meta, command);
+}

--- a/src/language_features/mod.rs
+++ b/src/language_features/mod.rs
@@ -1,3 +1,4 @@
+pub mod call_hierarchy;
 pub mod ccls;
 pub mod clangd;
 pub mod codeaction;

--- a/src/types.rs
+++ b/src/types.rs
@@ -156,6 +156,13 @@ pub struct PositionParams {
     pub position: KakounePosition,
 }
 
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CallHierarchyParams {
+    pub position: KakounePosition,
+    pub incoming_or_outgoing: bool,
+}
+
 #[derive(Clone, Copy, Deserialize, Debug)]
 #[serde(rename_all = "camelCase")]
 pub struct CodeActionsParams {

